dawn/src/dawn/SIR/SIR.h

```diff
@@ -149,15 +150,74 @@ struct StencilFunctionArg {
   CompareResult comparison(const sir::StencilFunctionArg& rhs) const;
 };
 
+class FieldDimensionImpl {
+public:
+  std::unique_ptr<FieldDimensionImpl> clone() const { return cloneImpl(); }
+  virtual ~FieldDimensionImpl() = default;
+
+private:
+  virtual std::unique_ptr<FieldDimensionImpl> cloneImpl() const = 0;
+};
+
+class CartesianFieldDimension : public FieldDimensionImpl {
+  bool maskI_, maskJ_, maskK_;
+  std::unique_ptr<FieldDimensionImpl> cloneImpl() const override {
+    return std::make_unique<CartesianFieldDimension>(maskI_, maskJ_, maskK_);
+  }
+
+public:
+  bool I() const { return maskI_; }
+  bool J() const { return maskJ_; }
+  bool K() const { return maskK_; }
+  ~CartesianFieldDimension() = default;
+  CartesianFieldDimension(std::array<bool, 3> mask)
+      : maskI_(mask[0]), maskJ_(mask[1]), maskK_(mask[2]) {}
+  CartesianFieldDimension(bool dimi, bool dimj, bool dimk)
+      : maskI_(dimi), maskJ_(dimj), maskK_(dimk) {}
+};
+
+class FieldDimension {
+  std::unique_ptr<FieldDimensionImpl> impl_;
+
+public:
+  FieldDimension() = delete;
+  ~FieldDimension() = default;
+  FieldDimension(dawn::ast::cartesian_, std::array<bool, 3> mask)
```

Comment:
Do we need both constructors?

Reply:
Actually I think both constructors should be removed.
@lukasm91 why do we need this constructor with tag in Offset/Extent?

Reply:
We resolved the second question. The first still remains. Maybe we should remove the second ctor.